src/room_member.cpp

```diff
@@ -0,0 +1,196 @@
+
+    peer->Send(&stream, HIGH_PRIORITY, RELIABLE_ORDERED, 0, server_address, false);
+}
+
+/**
+ * Sends a request to the server, asking for permission to join a room with the specified nickname and preferred mac.
+ * @params nickname The desired nickname.
+ * @params preferred_mac The preferred MAC address to use in the room, an all-zero MAC address tells the server to assign one for us.
+ */
+static void SendJoinRequest(RakNet::RakPeerInterface* peer, const std::string& nickname, const MacAddress& preferred_mac = NoPreferredMac) {
+    // TODO(Subv): Implement.
+}
+
+void RoomMember::ReceiveLoop() {
+    // Receive packets while the connection is open
+    while (IsConnected()) {
+        RakNet::Packet* packet = peer->Receive();
+        if (!packet)
+            continue;
+
+        switch (packet->data[0]) {
+        case ID_ROOM_CHAT:
+            //FIXME
+            break;
+        case ID_ROOM_WIFI_PACKET:
+            HandleWifiPackets(packet);
+            break;
+        case ID_ROOM_INFORMATION:
+            //FIXME: Update playerlist from packet
+            break;
+        case ID_ROOM_NAME_COLLISION:
+            state = State::NameCollision;
+            peer->CloseConnection(server_address, true);
+            break;
+        case ID_ROOM_MAC_COLLISION:
+            state = State::MacCollision;
+            peer->CloseConnection(server_address, true);
+            break;
+        case ID_ROOM_JOIN_SUCCESS:
+            // The join request was successful, we are now in the room.
+            // If we joined successfully, there must be at least one client in the room: us.
+            ASSERT_MSG(GetMemberInformation().size() > 0, "We have not yet received member information.");
+            state = State::Joined;
+            break;
+        case ID_DISCONNECTION_NOTIFICATION:
+            // Connection lost normally
+            state = State::Idle;
+            peer->CloseConnection(server_address, true);
+            return;
+        case ID_INCOMPATIBLE_PROTOCOL_VERSION:
+            state = State::WrongVersion;
+            peer->CloseConnection(server_address, true);
+            break;
+        case ID_CONNECTION_ATTEMPT_FAILED:
+            state = State::Error;
+            break;
+        case ID_NO_FREE_INCOMING_CONNECTIONS:
+            state = State::RoomFull;
+            peer->CloseConnection(server_address, true);
+            break;
+        case ID_CONNECTION_LOST:
+            // Couldn't deliver a reliable packet, the other system was abnormally terminated
+            state = State::LostConnection;
+            break;
+        case ID_CONNECTION_REQUEST_ACCEPTED:
+            // Update the server address with the address of the sender of this packet.
+            server_address = packet->systemAddress;
+            // TODO(Subv): Send a room join request to the server before marking us as joined.
+            state = State::Joined;
```

Comment:
The TODO comment only says to send the packet. What I'm saying is: wait for a response.
So the state should not be modified in this function at all.